packages/@react-stately/selection/src/useMultipleSelectionState.ts

```diff
@@ -39,10 +56,16 @@ export function useMultipleSelectionState(props: MultipleSelectionStateProps): M
   let [, setFocusedKey] = useState(null);
   let selectedKeysProp = useMemo(() => convertSelection(props.selectedKeys), [props.selectedKeys]);
   let defaultSelectedKeys = useMemo(() => convertSelection(props.defaultSelectedKeys, new Selection()), [props.defaultSelectedKeys]);
+  let onSelectionChange = (keys) => {
+    if (props.onSelectionChange && (alwaysFireOnSelection || !equalSets(keys, selectedKeys))) {
```

Comment:
Perhaps we should do this check as a wrapper around `setSelectedKeys` instead? That would also avoid unnecessary re-renders for uncontrolled components in addition to not firing the event.